### PR TITLE
Developer UX: add  developer settings submenu to SMP

### PR DESCRIPTION
**Pull request by sejas**
- Closes https://github.com/Automattic/dotcom-forge/issues/1626

## Proposed Changes

Create a submenu using `@wordpress/components` elements to display useful links for developers inside SMP.

* Create a new component `SubmenuPopover`
* Add Developer settings submenu item to SMP actions menu
* Add feature flag `dev/developer-ux` enabled only on development and wpcalypso.
* Add accessibility keyboard bindings to the new submenu.
* Add submenu items.
* Add `id` to Hosting configuration Cache Card
* Add `SubmenuPopover` storybook and Readme

The goal of this PR is just evaluating the submenu behaviour. Not each individual link.
That will come in separate issues:
- Closes https://github.com/Automattic/dotcom-forge/issues/1631
- Closes https://github.com/Automattic/dotcom-forge/issues/1632
- Closes https://github.com/Automattic/dotcom-forge/issues/1634
- Closes https://github.com/Automattic/dotcom-forge/issues/1636
- Closes https://github.com/Automattic/dotcom-forge/issues/1637
- https://github.com/Automattic/dotcom-forge/issues/1633
- https://github.com/Automattic/dotcom-forge/issues/1635
- https://github.com/Automattic/dotcom-forge/issues/1699

## Testing Instructions

* Go to `/sites`
* Click on the action menu of a business site.
* Observe a new entry `Developer settings`.
* Put the mouse hover.
* Observe a submenu with multiple items.

## Screencast

https://user-images.githubusercontent.com/779993/219529954-0cbd23d5-3ae9-4380-89e4-f6d002f497ec.mp4

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: packages/components/src/submenu-popover/index.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+ * Returns the anchor rect adding a given offsetTop.
+ */
+function useAnchorWithOffset( anchorRect: DOMRect | undefined, { offsetTop = 0 } = {} ) {
+	if ( ! anchorRect || offsetTop === 0 ) {
+		return anchorRect;
+	}
+	return new DOMRect( anchorRect.x, anchorRect.y + offsetTop, anchorRect.width, anchorRect.height );
+}
+
+export function useSubenuPopoverProps< T extends HTMLElement >(
```
Comment (wojtekn) [inline packages/components/src/submenu-popover/index.tsx]: It looks like a typo here, should it be `useSubmenuPopoverProps`?

File: client/sites-dashboard/components/sites-ellipsis-menu.tsx [hunk 2]
```diff
@@ -268,6 +372,7 @@ export const SitesEllipsisMenu = ( {
 				<SiteMenuGroup>
 					{ site.launch_status === 'unlaunched' && <LaunchItem { ...props } /> }
 					<SettingsItem { ...props } />
+					{ isEnabled( 'dev/developer-ux' ) && <DeveloperSettingsSubmenu { ...props } /> }
 					<ManagePluginsItem { ...props } />
 					{ showHosting && <HostingConfigItem { ...props } /> }
```
Comment (wojtekn) [inline client/sites-dashboard/components/sites-ellipsis-menu.tsx:377]: Should we remove Hosting Configuration link if we shod more granular links in the new submenu?

**Review by wojtekn — COMMENTED** [review]
It looks well and works for me for both mobile and desktop. I was able to use it using the keyboard too.

I've noticed one undesired behavior - when I open the submenu, hover it, and start scrolling the page, I'm able to disconnect the submenu from the menu and move it around:

Screen Capture on 2023-02-17 at 10-07-52

**Comment by javierarce** [comment]
Great work, @sejas! Except for the scroll thing Wojtek mentions, this works great. 

There is something small we should tweak (here or in a new ticket), which is the position of the icons inside the menu. We should match the one of the menus in Gutenberg, which is right-aligned.

image

**Comment by javierarce** [comment]
Oh, there's another thing: we should allow users to navigate in and out of the submenu with the left and right arrow keys (right to enter, left to exit). Right now, you can open the submenu by pressing Enter and exit it with the Esc key, but it's not very natural.

**Comment by sejas** [comment]
> It looks well and works for me for both mobile and desktop. I was able to use it using the keyboard too.
> 
> I've noticed one undesired behavior - when I open the submenu, hover it, and start scrolling the page, I'm able to disconnect the submenu from the menu and move it around:
> 

Thanks for reviewing it 🙌 .
I've fixed the scrolling issue here https://github.com/Automattic/wp-calypso/pull/73356/commits/9c8cb594fa3075a300cf4d2fb1073731e46dbed4

I also fixed the typo in the hook name and the icons alignment 👌 

@wojtekn , can you take another review? Thank you!!

File: packages/components/src/submenu-popover/index.tsx (hunk 1 above)
Comment (sejas) [reply packages/components/src/submenu-popover/index.tsx]: Fixed here: https://github.com/Automattic/wp-calypso/pull/73356/commits/35a63b1b9f8f0a2dd531234c6faeaf462321bd76
Thank you! 🙏

**Comment by sejas** [comment]
> Great work, @sejas! Except for the scroll thing Wojtek mentions, this works great.
> 
> There is something small we should tweak (here or in a new ticket), which is the position of the icons inside the menu. We should match the one of the menus in Gutenberg, which is right-aligned.
> 
> image

@javierarce , I've changed the icon alignment to the right 👌  https://github.com/Automattic/wp-calypso/pull/73356/commits/88b14d909b804f94fc4bc86146733bcce2e3b791
Thanks for pointing this out.

Screenshot 2023-02-17 at 11 51 11

File: client/sites-dashboard/components/sites-ellipsis-menu.tsx (hunk 2 above)
Comment (sejas) [reply client/sites-dashboard/components/sites-ellipsis-menu.tsx:377]: Good question. I had the same thought.
My reasoning for not doing it is that we would want to keep the `Hosting Configuration` link for Free sites.
From my point of view, it makes sense to remove it for atomic sites. I'll move the discussion to P2. 👌 
Thank you!

**Comment by wojtekn** [comment]
@sejas scrolling works well now, thanks! I've noticed two more things with how links work:
- "Privacy settings" link doesn't work well for unlaunched sites, as they don't have a "Privacy settings" block and have "Launch site" instead. Should we remove it, especially we display "Launch site" for them in the main menu?
- when I click "Performance settings", I land on the proper page, but it is scrolled somehow to half. The URL doesn't include any anchor. Does it work that way for you too?

**Comment by sejas** [comment]
> @sejas scrolling works well now, thanks! I've noticed two more things with how links work:
> 
> * "Privacy settings" link doesn't work well for unlaunched sites, as they don't have a "Privacy settings" block and have "Launch site" instead. Should we remove it, especially we display "Launch site" for them in the main menu?
> * when I click "Performance settings", I land on the proper page, but it is scrolled somehow to half. The URL doesn't include any anchor. Does it work that way for you too?

@wojtekn Let's land this PR and address those cases on separate issues:
- https://github.com/Automattic/dotcom-forge/issues/1635
- https://github.com/Automattic/dotcom-forge/issues/1699

I'll close only the issues linking to Hosting configuration sections 👌 .

**Review by wojtekn — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7819802

Hi @sejas, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Performance settings`
- `Github connection`
- `Developer settings`

Thank you in advance!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7819802).

File: client/sites-dashboard/utils.ts [hunk 3]
```diff
@@ -28,6 +28,10 @@ export const displaySiteUrl = ( siteUrl: string ) => {
 	return siteUrl.replace( 'https://', '' ).replace( 'http://', '' );
 };
 
+export function isCustomDomain( siteSlug: string ): boolean {
+	return ! siteSlug.endsWith( '.wordpress.com' ) && ! siteSlug.endsWith( '.wpcomstaging.com' );
```
Comment (obenland) [inline client/sites-dashboard/utils.ts:32]: @sejas I just came across a WSOD on the `/sites` page while scrolling down the list of sites. It looks like there can be cases where `siteSlug` is null and these checks lead to a type error:

image

Comment (sejas) [reply client/sites-dashboard/utils.ts:32]: @obenland , thanks for flagging the error.
I've created a PR that fixes that error https://github.com/Automattic/wp-calypso/pull/82345

**Pull request merged**